src/librustc_borrowck/borrowck/mir/elaborate_drops.rs

```diff
@@ -0,0 +1,1094 @@
+    is_cleanup: bool,
+
+    init_data: &'a InitializationData,
+    kind: ElaborateKind,
+
+    lvalue: &'a Lvalue<'tcx>,
+    path: MovePathIndex,
+    succ: BasicBlock,
+    unwind: Option<BasicBlock>
+}
+
+impl<'b, 'tcx> ElaborateDropsCtxt<'b, 'tcx> {
+    fn tcx(&self) -> TyCtxt<'b, 'tcx, 'tcx> { self.ctxt.0 }
+    fn mir(&self) -> &'b Mir<'tcx> { self.ctxt.1 }
+    fn move_data(&self) -> &'b MoveData<'tcx> { &self.ctxt.2 }
+
+    fn initialization_data_at(&self, loc: Location) -> InitializationData {
+        let mut data = InitializationData {
+            live: self.flow_inits.sets().on_entry_set_for(loc.block.index())
+                .to_owned(),
+            dead: self.flow_uninits.sets().on_entry_set_for(loc.block.index())
+                .to_owned(),
+        };
+        for stmt in 0..loc.index {
+            data.apply_location(self.ctxt.0, self.ctxt.1, &self.ctxt.2,
+                                Location { block: loc.block, index: stmt });
+        }
+        data
+    }
+
+    fn create_drop_flag(&mut self, index: MovePathIndex) {
+        let tcx = self.tcx();
+        let patch = &mut self.patch;
+        self.drop_flags.entry(index).or_insert_with(|| {
+            patch.new_temp(tcx.types.bool)
+        });
+    }
+
+    fn drop_flag(&mut self, index: MovePathIndex) -> Option<Lvalue<'tcx>> {
+        self.drop_flags.get(&index).map(|t| Lvalue::Temp(*t))
+    }
+
+    fn elaborate(mut self) -> (MirPatch<'tcx>, FnvHashMap<MovePathIndex, u32>) {
+        self.collect_drop_flags();
+        self.elaborate_drops();
+
+        self.drop_flags_on_init();
+        self.drop_flags_for_fn_rets();
+        self.drop_flags_for_args();
+        self.drop_flags_for_locs();
+
+        (self.patch, self.drop_flags)
+    }
+
+    fn path_needs_drop(&self, path: MovePathIndex) -> bool
+    {
+        match self.move_data().move_paths[path].content {
+            MovePathContent::Lvalue(ref lvalue) => {
+                let ty = self.mir().lvalue_ty(self.tcx(), lvalue)
+                    .to_ty(self.tcx());
+                debug!("path_needs_drop({:?}, {:?} : {:?})", path, lvalue, ty);
+
+                self.tcx().type_needs_drop_given_env(ty, &self.param_env)
+            }
+            _ => false
+        }
+    }
+
+    fn lvalue_concrete_strict_parent<'a>(&self, lv: &'a Lvalue<'tcx>)
+                                         -> Option<&'a Lvalue<'tcx>>
+    {
+        if let &Lvalue::Projection(ref data) = lv {
+            self.lvalue_concrete_parent(&data.base)
+        } else {
+            None
+        }
+    }
+
+    fn lvalue_concrete_parent<'a>(&self, lv: &'a Lvalue<'tcx>)
+                                  -> Option<&'a Lvalue<'tcx>>
+    {
+        let ty = self.mir().lvalue_ty(self.tcx(), lv)
+            .to_ty(self.tcx());
+        match ty.sty {
+            ty::TyArray(..) | ty::TySlice(..) | ty::TyRef(..) | ty::TyRawPtr(..) => {
+                Some(lv)
+            }
+            _ => match lv {
```

Comment:
Personally, I think I would have spelled out the other ty variants here, but I won't make you do it :)